tab3 update

bhlieberman opened the pull request.

bhlieberman committed:
initial commit on dev branch

bhlieberman committed:
refactored `00_download` to use functions and housekeeping on both

bhlieberman committed:
DVC initialized and run for first time

bhlieberman committed:
logging added and now renaming columns to remove #

bhlieberman committed:
dvc repro build latest

bhlieberman committed:
specify that pipe-separated columns should be interpreted as strings to avoid undesired behavior (silent failure) when reading TSV files

bhlieberman committed:
add logging

bhlieberman committed:
latest dvc run

bhlieberman committed:
use a `ThreadPoolExecutor` to concurrently download files

bhlieberman committed:
synchronize branch histories

zmughal commented on stages/00_download.py:
```diff
@@ -0,0 +1,68 @@
+    return soup
+
+
+tab3_files = re.compile("https://.*/Download/(?!.*OSPREY).*\.tab3.zip$")
```
Looks good so far, but we need more than the `tab3` format. These are the files that we should get:

```
BIOGRID-ALL-4.4.232.tab3.zip
BIOGRID-CHEMICALS-4.4.232.chemtab.zip
BIOGRID-CORONAVIRUS-4.4.232.tab3.zip
BIOGRID-IDENTIFIERS-4.4.232.tab.zip
BIOGRID-MV-Physical-4.4.232.tab3.zip
BIOGRID-ORGANISM-4.4.232.tab3.zip
BIOGRID-PROJECT-autophagy_project-4.4.232.zip
BIOGRID-PROJECT-covid19_coronavirus_project-4.4.232.zip
BIOGRID-PROJECT-fanconi_anemia_project-4.4.232.zip
BIOGRID-PROJECT-glioblastoma_project-4.4.232.zip
BIOGRID-PROJECT-kinome_project_sc-4.4.232.zip
BIOGRID-PROJECT-ups_project-4.4.232.zip
BIOGRID-PTMS-4.4.232.ptm.zip
BIOGRID-SYSTEM-4.4.232.tab3.zip
```

diff with the current download:
```diff
--- b	2024-04-18 13:57:54.112292341 -0400
+++ a	2024-04-18 13:55:49.854351188 -0400
@@ -1,5 +1,14 @@
 BIOGRID-ALL-4.4.232.tab3.zip
+BIOGRID-CHEMICALS-4.4.232.chemtab.zip
 BIOGRID-CORONAVIRUS-4.4.232.tab3.zip
+BIOGRID-IDENTIFIERS-4.4.232.tab.zip
 BIOGRID-MV-Physical-4.4.232.tab3.zip
 BIOGRID-ORGANISM-4.4.232.tab3.zip
+BIOGRID-PROJECT-autophagy_project-4.4.232.zip
+BIOGRID-PROJECT-covid19_coronavirus_project-4.4.232.zip
+BIOGRID-PROJECT-fanconi_anemia_project-4.4.232.zip
+BIOGRID-PROJECT-glioblastoma_project-4.4.232.zip
+BIOGRID-PROJECT-kinome_project_sc-4.4.232.zip
+BIOGRID-PROJECT-ups_project-4.4.232.zip
+BIOGRID-PTMS-4.4.232.ptm.zip
 BIOGRID-SYSTEM-4.4.232.tab3.zip
```

zmughal commented on stages/01_process.py:
```diff
@@ -0,0 +1,103 @@
+    with open(file_path, "rb") as zip:
+        try:
+            zip_root = zipfile.ZipFile(zip)
+            zip_root.extractall(Path("unzip"))
```
Can we unzip to a directory with the same base name as the `.zip` file? That way the files will be grouped together in directories under the `brick/` directory as well.

zmughal reviewed (changes requested):
Looks good. Couple more changes to what is downloaded and the directory structure and I think this will be done.

We should also add a `flake.nix` with the requirements. Example [here](https://github.com/biobricks-ai/mirbase/pull/1/files?file-filters%5B%5D=.nix&show-deleted-files=true&show-viewed-files=true).

bhlieberman committed:
`flake.nix` added

bhlieberman committed:
scripts updated to download all files and create appropriate subdirectories

bhlieberman committed:
latest DVC run

zmughal commented on stages/00_download.py:
```diff
@@ -27,16 +27,15 @@ def get_biogrid_html():
 
-
-tab3_files = re.compile("https://.*/Download/(?!.*OSPREY).*\.tab3.zip$")
+REGEX = "^https://.*/Download/.*/BIOGRID-((?!OSPREY).*\.*(tab3|chemtab|ptm|-4.4.232)|(IDENTIFIERS-4.4.232.tab)).zip$"
```
Is there a way we can extract out the version number `4.4.232` to a file? I think a YAML file that just contains `version: 4.4.232` would work well. The scripts that use it can depend on it in `dvc.yaml` (this one and the other).

Also, this is a case for using `re.escape()` because the `.` will be treated a the regex meta-character.

zmughal commented on stages/01_process.py:
```diff
@@ -0,0 +1,119 @@
+                     low_memory=False,
+                     on_bad_lines='skip')
+    if df is not None:
+        df.rename(columns={"#BioGRID Interaction ID": "BioGRID Interaction ID"},
```
Also:

```
"#BIOGRID ID"
"#BioGRID Chemical Interaction ID"
"#PTM ID"
```

zmughal commented on stages/01_process.py:
```diff
@@ -0,0 +1,119 @@
+            logger.debug("Problem unzipping file", exc_info=1)
+
+
+DTYPES = {k: pd.StringDtype(storage="pyarrow") for k in
```
These can also be treated as strings (some of the are integer IDs, but treating them as strings is fine):

```
"Entrez Gene Interactor A"
"Entrez Gene Interactor B"
"Related BioGRID Gene ID"
"Related Entrez Gene ID"
"Related Organism ID"
"Synonyms Interactor A"
"Synonyms Interactor B"
"Systematic Name Interactor A"
"Systematic Name Interactor B"
```

zmughal commented on flake.nix:
```diff
@@ -0,0 +1,20 @@
+        devShells.default = dev-shell.devShells.${system}.default.overrideAttrs
+          (oldAttrs: {
+            buildInputs = oldAttrs.buildInputs ++ [
+              (python3.withPackages (ps: with ps; [ bs4 numpy pandas pyarrow requests ]))
```
Nix package for bs4 is called `beautifulsoup4`.

zmughal reviewed (changes requested):
See review comments.

bhlieberman committed:
split unzipping logic out into separate stage

bhlieberman committed:
corrected nix package name

bhlieberman committed:
added dependency on pyyaml and using it to track biogrid release version

bhlieberman committed:
some changes to parquet logic and file renamed to `02_process`

bhlieberman committed:
stop tracking version.yaml

bhlieberman committed:
latest changes and run

zmughal committed:
Minor regex escaping fixes

zmughal committed:
Add more column-to-type mappings

Analysed with:

```shell
duckdb -c '.mode csv' -c "$(cat <<EOF
    SELECT
      name,
      list(DISTINCT type) AS types,
      list(file_name)[1:3] AS files
    FROM parquet_schema('brick/**/*.parquet')
    WHERE
    type NOT IN ('BYTE_ARRAY', 'INT64')
    AND
    file_name NOT LIKE '%PROJECT%'
    GROUP BY name
    ORDER BY name
    ;
EOF
)";
```

zmughal committed:
Add /download to .gitignore

zmughal committed:
Update flake.lock

zmughal committed:
Simplify unzip script

zmughal reviewed (approved):
Made a couple changes. Looks good!
